### PR TITLE
Standardized behavior of Batch.cat and misc code refactor

### LINKED ISSUE #131: Remove some useless kwargs or not
For some functions (like ``Collector.__init__``), @Trinkle23897 keep ``**kwargs`` arguments for api stability, so that adding a new argument with default values can work with older versions of tianshou.

However, I tend to remove these unused ``kwargs`` because they may hide some bugs and make it difficult for IDE to report unexpected arguments.

We need some discussions to determine what to do next. Suggestions are welcome especially by @duburcqa . Must we keep the ability of old versions of tianshou to work with code written for new versions of tianshou?

### LINKED ISSUE #138: The length of empty Batch
Currently, ``len(Batch())`` will raise an error.

The numpy behavior is:
```python
a = np.array([])
len(a) # 0
b = np.array(3)
len(b) # exception
```

So, len of an empty Batch should be 0. len of a Batch with scalars should raise an error.

Besides, if we want to support ``cat`` with empty Batches, we should also make ``len(Batch()) == 0``, otherwise we cannot tell ``Batch()`` from ``Batch(a=1)``, where ``Batch()`` can be used in ``cat`` but the scalar ``Batch(a=1)`` cannot.

@Trinkle23897  @duburcqa

### LINKED ISSUE #139: Standardize the behavior of Batch aggregation (stack/cat) when dealing with reserved keys
Currently, we use ``Batch()`` to indicate that a key reserves the place in Batch and it will have value later on. It works fine in most cases, but now we have noticed that it is problematic. 

The critic problem is: how to treat ``Batch(c=Batch())``? There are two opinions:

- ``Batch(c=Batch())`` means hierarchical key reservation, and should be treated as empty. But the problem is, we want to enable the concatenation of Batches with other empty Batches, which is seemingly natural. But considering ``Batch.cat([Batch(a=np.random.rand(3, 4)), Batch(a=Batch(c=Batch()))])``, we would want to treat ``Batch(c=Batch())`` as non-empty.

- ``Batch(c=Batch())`` is not empty, and there is no need to support hierarchical key reservation. This makes the implementation straightforward, but does not support hierarchical key reservation. Unfortunately, there are some use cases of hierarchical key reservation in Tianshou.

I think the critical problem is how to reserve keys and support hierarchical key reservation.

@Trinkle23897  @duburcqa

### PR DESCRIPTION
**Pull request by youkaichao**
Misc code refactor and add Batch.condense

These changes are picked from #122 to be an independent PR.

### CODE REVIEW

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: Why is this useful? Plus I don't find this name very informative.

Comment (youkaichao) [reply tianshou/data/batch.py]: When ``cat`` does not support to concatenate a list of batches with some fields filled by empty Batch() (the default behavior of ``numpy.concatenate``), but we need this feature for multi-agent RL. To be consistent with ``numpy``, we can remove empty fields first, and then ``cat`` them.

The functionality is to remove fields with empty Batch, it is ok if you have another name.

Comment (duburcqa) [reply tianshou/data/batch.py]: Why not to support empty field in `cat` instead ? Using the "key reservation" paradigm (empty Batch are not regular keys). It is a feature that it not part of numpy.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py:415]: I noticed this yesterday but I didn't care to fix it x) Nice thing that you did it !

File: tianshou/policy/base.py
Comment (duburcqa) [inline tianshou/policy/base.py:154]: typo I guess.

File: tianshou/policy/base.py
Comment (duburcqa) [inline tianshou/policy/base.py]: rew = rew ?

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: I tried, but that's very annoying. Take a code snippet for example:
```python
b1 = Batch(a=np.random.rand(3, 4), common=Batch(c=np.random.rand(3, 5)))
b2 = Batch(b=torch.rand(4, 3), common=Batch(c=np.random.rand(4, 5)))
test = Batch.cat([b1, b2])
```
For partial key 'a', 'a' is not a key of b2, so it is perfectly ok to fill zero in b2. If we support b1.a is Batch() in ``cat``, what is the output ``test.a`` supposed to be? Furthermore, if both b1.a and b2.a is Batch(), what is the output ``test.a`` supposed to be?

If these are made clear, I'm happy to support ``cat`` with empty Batch().

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py:415]: The exception comes up when I write MARL scripts, so I fixed it by the way.

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: For shared keys, it is also very annoying. Currently, we are passing ``cat`` to ``np.concatenate`` for shared keys. The support for ``cat`` with Batch() would make the code much complicated, essentially we need to duplicate the code of dealing with partial keys.

Comment (youkaichao) [reply tianshou/data/batch.py]: One possible solution would be to re-define shared keys, like ``keys_map = [set(k for k, v in batch.items() if not (isinstance(v, Batch) and v.is_empty())) for batch in batches]``, omitting keys with empty Batch() when calculating shared keys.

Comment (Trinkle23897) [reply tianshou/data/batch.py]: My opinion is

> If we support `b1.a` is Batch() in `cat`, what is the output `test.a` supposed to be? 

`Batch()`, since `b1.a` is `Batch()` and `b2.a` does not exist.

> Furthermore, if both `b1.a` and `b2.a` is Batch(), what is the output `test.a` supposed to be?

`Batch()`.

But it should define that `len(Batch()) == 0` I guess?

Comment (duburcqa) [reply tianshou/data/batch.py]: > One possible solution would be to re-define shared keys, like keys_map = [set(k for k, v in batch.items() if not (isinstance(v, Batch) and v.is_empty())) for batch in batches], omitting keys with empty Batch() when calculating shared keys.

Exactly.  You just need to compute 3 set of keys: 
- shared keys -> Not empty for all batches
- partial keys -> Empty for at least one of the batches AND not empty for  at least one of the batches
- reserved keys -> Empty for at least one of the batches

Comment (duburcqa) [reply tianshou/data/batch.py]: > If we support b1.a is Batch() in cat, what is the output test.a supposed to be?

`b1.a` is `np.random.rand(3, 4)`, and `b2.a` is `Batch()`, so it is the (newly refined) "partial keys" case.

> if both b1.a and b2.a is Batch(), what is the output test.a supposed to be?

if both `b1.a` and `b2.a` are `Batch()`, then it is the (newly refined) "reserved keys" case. So you should only reserve the key in the concatenated batch.

Comment (duburcqa) [reply tianshou/data/batch.py]: I think the 3 cases are pretty clear, natural, and quite straightforward to implement. You just need to refine the computation of "shared keys" and "partial keys", and add a new 5 lines of code "reserved keys" handling.

Comment (youkaichao) [reply tianshou/data/batch.py]: > > One possible solution would be to re-define shared keys, like keys_map = [set(k for k, v in batch.items() if not (isinstance(v, Batch) and v.is_empty())) for batch in batches], omitting keys with empty Batch() when calculating shared keys.
> 
> Exactly. You just need to compute 3 set of keys:
> 
> * shared keys -> Not empty for all batches
> * partial keys -> Empty for at least one of the batches AND not empty for  at least one of the batches
> * reserved keys -> Empty for at least one of the batches

Okay, reserved keys are treated the same as partial keys.

@Trinkle23897 
> But it should define that len(Batch()) == 0 I guess?

No, we can first filter out empty batches by ``batches = [b for b in batches if len(b.__dict__)]``.

Comment (youkaichao) [reply tianshou/data/batch.py]: > I think the 3 cases are pretty clear, natural, and quite straightforward to implement. You just need to refine the computation of "shared keys" and "partial keys", and add a new 5 lines of code "reserved keys" handling.

What about ``stack``? The same semantic meaning may apply, too.

File: tianshou/policy/base.py
Comment (youkaichao) [reply tianshou/policy/base.py:154]: A trailing comma is okay, though. Future modifications then will not include this comma.

File: tianshou/data/batch.py
Comment (duburcqa) [reply tianshou/data/batch.py]: Yes it should.

**Comment by youkaichao** [comment]
Currently, ``cat`` only supports partially empty Batch, that is, a non-empty Batch but has some fields with empty Batch.

To enable ``cat`` with empty Batch like ``Batch(a=Batch(), b=Batch(c=Batch()))``, we have to know its length, which should be resolved after #138 .

**Comment by duburcqa** [comment]
> To enable cat with empty Batch like Batch(a=Batch(), b=Batch(c=Batch())), we have to know its length, which should be resolved after #138 .

`is_empty` is provided in such case. No need to implement change or implement anything else.

**Comment by youkaichao** [comment]
```python
Batch.cat([Batch(a=np.random.rand(3, 4)), Batch(b=Batch(), a=Batch(c=Batch()))])
```
@duburcqa What about this? Maybe this should cause an error? Then we should distinguish between Batch() and Batch(c=Batch()), where the former is used to reserve keys but the latter should not be treated as empty.

This raises a natural question: is it ok to use Batch() to reserve keys? Or maybe we should explicitly reserve the keys? (e.g. have a field named ``self.reserved_keys``).

Or, must we reserve any keys? Maybe we can reserve no keys, return a special token to indicate the absence of a key, and let the caller to deal with it.

One more issue: ``None`` is also a valid object (``Batch(a=None)`` works just fine), but we are using ``b.get(k, None) is None`` to judge the existence of a key.

**Comment by duburcqa** [comment]
> @duburcqa What about this? 

`Batch(c=Batch())` is not an empty Batch -> `a=Batch(c=Batch())` is not a key reservation -> the shape of `Batch(b=Batch(), a=Batch(c=Batch()))` is not compatible with the shape of `Batch(a=np.random.rand(3, 4))` -> it should raise an exception. I see no ambiguity here.

> This raises a natural question: is it ok to use Batch() to reserve keys? Or maybe we should explicitly reserve the keys? (e.g. have a field named self.reserved_keys).
Or, must we reserve any keys? Maybe we can reserve no keys, return a special token to indicate the absence of a key, and let the caller to deal with it.
One more issue: None is also a valid object (Batch(a=None) works just fine), but we are using b.get(k, None) is None to judge the existence of a key.

I dont know. But those design issues should be discussed in another dedicated issue + PR.

**Comment by youkaichao** [comment]
> I see no ambiguity here.

Currently, ``Batch(c=Batch())`` is empty. And maybe someone would want to use ``Batch(b=Batch(), a=Batch(c=Batch()))`` for hierarchical key reservation?

**Comment by duburcqa** [comment]
> Currently, Batch(c=Batch()) is empty. 

Yes, but I don't think it should (at least for this PR).

> And maybe someone would want to use Batch(b=Batch(), a=Batch(c=Batch())) for hierarchical key reservation?

Maybe, but I think it is another issue since it requires changing of paradigm. But it my opinion, it is irrelevant to support hierarchical key reservation. I may be wrong.

**Comment by Trinkle23897** [comment]
Hierarchical key reservation is a must for the current collector implementation, since it will create something like self.data is a `Batch(obs=..., ..., policy=Batch(_state=Batch()), ...)`

**Comment by youkaichao** [comment]
> Currently, Batch(c=Batch()) is empty.

Now we support ``is_empty(recursive=bool)`` to identify either ``Batch()`` or ``Batch(c=Batch())``.

> Hierarchical key reservation is a must for the current collector implementation

Hierarchical key reservation is stays.

With 6d2cda6 , now ``Batch.cat_`` behaves exactly as described [here](https://github.com/thu-ml/tianshou/issues/139#issuecomment-659092494).

**Comment by youkaichao** [comment]
> One more issue: None is also a valid object (Batch(a=None) works just fine), but we are using b.get(k, None) is None to judge the existence of a key.

2541bec fixes it.

**Comment by duburcqa** [comment]
> Now we support is_empty(recursive=bool) to identify either Batch() or Batch(c=Batch()).

Torch is using  `recurse` not `recursive`.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py:27]: Nice catch !

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: Reserving keys `a=Batch(a=Batch())` should not be necessary.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: Never use `Exception`. Too generic. `ValueError` is more appropriate.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: Why not to use only the `try...catch` black ?

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: I don't like such the idea of using hint.

Comment (youkaichao) [reply tianshou/data/batch.py]: A better solution is welcome. But it seems unavoidable for the case I give:

```python
x=Batch(a=Batch(a=np.random.randn(3, 4)), b=np.random.randn(3, 4))
y = Batch(a=Batch(a=Batch()), b=np.random.randn(3, 4))
Batch.cat([x, y])
```

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: This is hierarchical key reservation, and is indeed required for tianshou:

> Hierarchical key reservation is a must for the current collector implementation, since it will create something like self.data is a `Batch(obs=..., ..., policy=Batch(_state=Batch()), ...)`

File: tianshou/data/batch.py
Comment (duburcqa) [reply tianshou/data/batch.py]: Yes, this issue was already discussed in the past. It is possible. You need to do something like this:

```
def cat(batches):
    def cat_impl(batches, lens):
        [...] # actual stack implementation

    lens = [...] # compute len here
    return cat_impl(batches, lens)
```

This way, it hides the implementation of `cat` that is used internally, because the `lens` parameter should not appears at top level in my opinion. It is quite easy to adapt your code to use this pattern, `cat_impl` is your current implementation of `cat`.

Comment (youkaichao) [reply tianshou/data/batch.py]: Agree that ``lens `` should not be exposed to users. But creating a nested function would make the indentation too deep. I changed it into an internal function ``__cat``.

File: tianshou/data/batch.py
Comment (duburcqa) [reply tianshou/data/batch.py]: That's not right. It should be possible to do this:
```
>>> Batch.stack([Batch(a=np.array([1])), Batch(a=np.array([2]), b=np.array([3]))])
Batch(
    a: array([[1],
              [2]]),
    b: array([[0],
              [3]]),
)
```

This is very likely to append in practice (and it happens in some project of my company). For example, concatenating environment `info`, the keys of the last step (when done flag is True) can be completely different, or a least some new ones can be present, for instance in the case of a terminal reward.

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: Can you elaborate on that? I do not get it.

File: tianshou/data/batch.py
Comment (duburcqa) [reply tianshou/data/batch.py]: By the way, take advantage of this structure to make `lens` argument required and not optional in `cat_impl`, and move the computation of the initial value outside `cat_impl`. It will make the complete implementation clearer and simpler.

Comment (duburcqa) [reply tianshou/data/batch.py]: > But creating a nested function would make the indentation too deep. I changed it into an internal function __cat.

As you wish. It is just an implementation detail.

File: tianshou/data/batch.py
Comment (duburcqa) [reply tianshou/data/batch.py]: I was just wonderting if this line is necessary since you are using recursive `is_empty` in the try/catch block.

Comment (youkaichao) [reply tianshou/data/batch.py]: Oh, this is necessary. ``is_empty()`` and ``is_empty(recursive=True)`` mean different things and have to be dealt with differently.

For ``is_empty() == True``, this is ``Batch()``, and will be ignored because this is a top-level cat.

For ``is_empty(recursive=True) == True``, this is nested empty Batch but not Batch(), and we have to treat it as length zero and keep it.

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: > That's not right. It should be possible to do this:

I don't get it. It is already supported now, ``Batch.stack([Batch(a=np.array([1])), Batch(a=np.array([2]), b=np.array([3]))])`` works as expected.

Comment (duburcqa) [reply tianshou/data/batch.py]: So the doc is wrong. Discussed in the related issue.

File: tianshou/data/batch.py
Comment (duburcqa) [reply tianshou/data/batch.py]: OK nice.

Comment (duburcqa) [reply tianshou/data/batch.py]: Maybe add a small comment at the end of those lines to make it clear.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py:545]: - take advantage of the new code architecture (cad combining `cat_` and `__cat`) to make `lens` argument required and not optional
- move the computation of the initial value of `lens` in `cat_` itself. 

It will make the complete implementation clearer and simpler.

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: Solved in #139 .

Comment (duburcqa) [reply tianshou/data/batch.py]: Mark resolved after fixing doc.

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py:545]: Fixed in 6287326 .

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: The doc is this [comment](https://github.com/thu-ml/tianshou/issues/139#issuecomment-659092494) is already fixed.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: `without ``lens``
        as a hint, when we concatenate x.a and y.a, we would not be able to
        know how to pad y.a. So we first compute the variable ``lens`` to the
        internal method ``Batch.__cat`` a hint.`

Not relevant here.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: I don'k like much this example since it implies that `a=Batch(a=Batch())` has to be specified. Perhaps add an extra field to x: `x=Batch(a=Batch(a=np.random.randn(3, 4)), b=np.random.randn(3, 4), c=np.random.randn(3, 2))`

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: Indeed this should not appear in the doc string of ``cat_``, which can be seen by users. Moved to the doc string of ``__cat`` in f64faf5 .

File: tianshou/data/batch.py
Comment (duburcqa) [reply tianshou/data/batch.py]: Sure, but it has to be fixed in this PR, line 137.

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: I think it is okay. It is a corner case that requires some thinking. 

``x=Batch(a=Batch(a=np.random.randn(3, 4)), b=np.random.randn(3, 4), c=np.random.randn(3, 2))`` is straightforward and actually previous code works for it, by treating 'c' as partial keys.

So we had better keep it to let future maintainers (including us) understand the necessity of ``batch_holder.__cat(v, lens=lens)``.

Comment (duburcqa) [reply tianshou/data/batch.py]: As you wish.

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: Oh, my bad. Fixed in b795493 .

**Comment by youkaichao** [comment]
> Torch is using `recurse` not `recursive`.

Fixed in 02f9d1a .

**Review by duburcqa — DISMISSED** [review]

**Comment by duburcqa** [comment]
Nice job ! Very high quality PR again :+1: @Trinkle23897 Ready to be merged.

NB: @Trinkle23897 you should not allow every collaborators to merge PR, or clearly state if you agree if I merge some PRs.

**Comment by Trinkle23897** [comment]
> NB: @Trinkle23897 you should not allow every collaborator to merge PR, or clearly state if you agree if I merge some PRs.

2020-07-16 16-34-37 的屏幕截图

I don't know how to set it since the minimum role for approval is "write". I open the " At least 1 approving review by reviewers with write access." choice. Maybe set it to 2 is better for our three?

**Comment by duburcqa** [comment]
> I don't know how to set it since the minimum role for approval is "write".

Settings -> branches -> branch protected rules -> master -> proected matching branches -> Require review from Code Owners + Restrict who can dismiss pull request reviews 

> Maybe set it to 2 is better for our three?

Why not

File: tianshou/data/batch.py
Comment (Trinkle23897) [inline tianshou/data/batch.py]:
```python
In [27]: b=Batch(a=np.zeros([3,4]),b=np.zeros([5,4]),c={})

In [28]: c=Batch(a=np.zeros([4,4]),b=np.zeros([5,4]),c=np.zeros([1,4]))

In [29]: Batch.cat([b,c]).c.shape
Out[29]: (4, 4)
```
Please specify this undefined behavior in the documentation or unittest. (choose the smallest size)

File: tianshou/data/batch.py
Comment (Trinkle23897) [inline tianshou/data/batch.py]: Should add 4 space before `>>>` and another space after `>>>`:
```python
    >>> Batch().is_empty()
    True
```
I think it is good to render the documentation and take a look before the commit.

File: test/base/test_batch.py
Comment (youkaichao) [reply test/base/test_batch.py:121]: ``set`` object is a normal object, and this is expected. In fact, this feature is used in multi-agent RL to store legal actions (store list of sets as np.ndarray with np.object type).

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: ``Batch.cat_`` is an example.

Comment (youkaichao) [reply tianshou/data/batch.py]: This should go into the doc string of ``len``.

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: Maybe we can add the doc render as a test, too, just like pep8 check.

Comment (Trinkle23897) [reply tianshou/data/batch.py]: Agree

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: Examples added in doc string.

**Review by Trinkle23897 — APPROVED** [review]

**Comment by Trinkle23897** [comment]
@duburcqa I've changed the number to 2.

**Review by duburcqa — APPROVED** [review]

**Pull request merged**
